src/examples/design/icons/SvgSearch.tsx

```diff
@@ -29,85 +31,134 @@ const StyledSvgWrapper = styled.div<{ isAnswerBot?: boolean }>`
   searchEnabled?: boolean;
   data: {
-    edges: [
-      { node: { name: string; relativeDirectory: string; childGardenSvg: { content: string } } }
-    ];
+    edges: ISvgNodeProps[];
   };
-}> = ({ data, searchEnabled }) => {
-  const [inputValue, setInputValue] = useState('');
-  const [debouncedInputValue, setDebouncedInputValue] = useState(inputValue);
-
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  const updatedDebouncedInputValue = useCallback(
-    debounce((value: string) => {
-      setDebouncedInputValue(value);
-    }, 300),
-    []
-  );
+}
 
-  useEffect(() => {
-    updatedDebouncedInputValue(inputValue);
-  }, [inputValue, updatedDebouncedInputValue]);
-
-  const StyledCol = styled(Col).attrs({ forwardedAs: 'li' })``;
+const Icon = (edge: ISvgNodeProps) => {
+  return (
+    <StyledCol lg={3} md={4} xs={6}>
+      <StyledIconWrapper>
+        <StyledSvgWrapper
+          isAnswerBot={edge.node.name === 'answer-bot'}
+          // eslint-disable-next-line @typescript-eslint/naming-convention
+          dangerouslySetInnerHTML={{ __html: edge.node.childGardenSvg.content }}
+        />
+        <Code size="small" title={edge.node.name}>
+          {edge.node.name}
+        </Code>
+      </StyledIconWrapper>
+    </StyledCol>
+  );
+};
 
-  const StyledRow = styled(Row).attrs({ forwardedAs: 'ul' })``;
+export const SvgSearch: React.FC<ISvgSearchProps> = ({ data, searchEnabled }) => {
+  const [inputValue, setInputValue] = useState('');
+  const [collapsed, setCollapsed] = useState(true);
 
   const icons = useMemo(() => {
     return data.edges
       .filter(edge => {
-        const formattedSearchValue = debouncedInputValue.trim().toLowerCase();
+        const formattedSearchValue = inputValue.trim().toLowerCase();
 
+        // Returns every icons, since the search value is empty.
         if (formattedSearchValue.length === 0) {
           return true;
         }
 
+        if (formattedSearchValue.length) {
+          setCollapsed(false);
+        }
+
         return edge.node.name.trim().toLowerCase().includes(formattedSearchValue);
       })
-      .map(edge => (
-        <StyledCol key={edge.node.name} lg={3} md={4} xs={6}>
-          <StyledIconWrapper>
-            <StyledSvgWrapper
-              isAnswerBot={edge.node.name === 'answer-bot'}
-              // eslint-disable-next-line @typescript-eslint/naming-convention
-              dangerouslySetInnerHTML={{ __html: edge.node.childGardenSvg.content }}
-            />
-            <Code size="small" title={edge.node.name}>
-              {edge.node.name}
-            </Code>
-          </StyledIconWrapper>
-        </StyledCol>
-      ));
-  }, [data, debouncedInputValue, StyledCol]);
+      .map(edge => {
+        return <Icon key={edge.node.name} {...edge} />;
+      });
+  }, [data.edges, inputValue]);
+
+  const onInputChange = debounce((event: React.ChangeEvent<HTMLInputElement>) => {
+    setInputValue(event.target.value);
+  }, 300);
```

Comment:
The debounced function is unnecessarily created each time a user types (state change). I think this was prevented previously by using `useCallback`. Another way to address this is to store the debounced function inside a `ref`. To ensure there aren't memory leaks, the component should clean up the debounced function:

```jsx
const Comp = () => {
  const [state, setState] = useState();
  const debouncedFn = useRef();

  useEffect(() => {
    debouncedFn.current = debounce(setState, 250);
    
    return () => {
      debouncedFn.current.cancel();
    }
  }, []);

  // ...
}

```

Reply:
Ah, interesting. Does using `useCallback` also resolve the issue with debounce having memory leaks?

Reply:
Yeah `useCallback` and `useMemo` should also work because they cache the debounced function. A `useEffect` is still necessary to perform clean up via `debouncedFn.cancel()`.

> The debounced function comes with a cancel method to cancel delayed func invocations